### PR TITLE
support minitest after_run

### PR DESCRIPTION
**Pull request by grosser**
Currently rails does not call after_run hooks, which means tests that use them work differently when run with `ruby test.rb` vs `rails test test.rb`

see https://github.com/seattlerb/minitest/blob/f9605387e4af7d657921a83aaf0ae364f6d26a57/lib/minitest.rb#L51-L65

### CODE REVIEW

File: railties/lib/rails/commands/test.rb [hunk 1]
```diff
@@ -1,9 +1,11 @@
 require "rails/test_unit/minitest_plugin"
 
 if defined?(ENGINE_ROOT)
-  $: << File.expand_path("test", ENGINE_ROOT)
+  $LOAD_PATH << File.expand_path("test", ENGINE_ROOT)
```
Comment (grosser) [inline railties/lib/rails/commands/test.rb]: drive-by cleanup ... more readable

File: railties/test/application/test_test.rb [hunk 2]
```diff
@@ -289,7 +321,7 @@ class UserTest < ActiveSupport::TestCase
       def assert_unsuccessful_run(name, message)
         result = run_test_file(name)
         assert_not_equal 0, $?.to_i
-        assert result.include?(message)
+        assert_includes result, message
```
Comment (grosser) [inline railties/test/application/test_test.rb]: drive by cleanup: assert x.include? does not give a nice error message when it fails

File: railties/test/application/test_test.rb [hunk 3]
```diff
@@ -26,6 +26,38 @@ def test_truth
...
+            assert true
+          end
+        end
+      RUBY
+
+      result = assert_successful_test_run "unit/foo_test.rb"
+      assert_equal ["HELLO", "WORLD"], result.scan(/HELLO|WORLD/) # only once and in correct order
+    end
+
+    test "simple failed test" do
```
Comment (grosser) [inline railties/test/application/test_test.rb]: little bit of extra test coverage since that was never tested in isolation

**Comment by grosser** [comment]
FYI monkey-patch until this is fixed:

```Ruby
            # rails test does not trigger after_run and rake does not work with at_exit
            # https://github.com/rails/rails/pull/26515
            callback = -> { File.delete(new_path) if File.symlink?(new_path) }
            if Minitest.respond_to?(:run_with_rails_extension) && Minitest.run_with_rails_extension
              at_exit(&callback)
            else
              Minitest.after_run(&callback)
            end
```

**Comment by kaspth** [comment]
Not liking how much we have to reimplement Minitest because we call `run` instead of `autorun`. I'd rather we explored if we could use `autorun` instead.

**Comment by grosser** [comment]
hmm not a big fan either, but would rather fix the bug first and refactor after :)

**Comment by kaspth** [comment]
Wouldn't consider investigating using autorun instead of exit + run directly a refactoring when it would fix this issue. Per my understanding it would also fix: https://github.com/rails/rails/issues/25048#issuecomment-247154483

cc @arthurnn

**Comment by grosser** [comment]
updated to use autorun ... my app worked fine with that change / tests are happy too ... but not sure if that will have any other side effects ...

**Comment by grosser** [comment]
... it seems to break [single_cov](https://github.com/grosser/single_cov) ... by changing the order in which at_exit hooks are getting called ... possibly also simple_cov ... 

should solve https://github.com/rails/rails/issues/25048 though ... maybe it should not call run/autorun at all and leave that to the test ... even less messing around with stuff and execution order is the same no matter if rake/ruby/rails-test is used ?

**Comment by grosser** [comment]
this is a lot of magic going on here ... so I'd prefer to not rewrite/check everything since chances are it's there for a reason ... solve this by adding after_run and solve the other by checking `at_exit_installed` and we get a tiny and safe patch ... rewriting can be done in another PR and will have good tests to fall back on ...

File: railties/lib/rails/commands/test.rb [hunk 4]
```diff
@@ -1,9 +1,9 @@
...
 else
-  $: << File.expand_path("../../test", APP_PATH)
+  $LOAD_PATH << File.expand_path("../../test", APP_PATH)
 end
 
-exit Minitest.run(ARGV)
+Minitest.autorun
```
Comment (arthurnn) [inline railties/lib/rails/commands/test.rb]: I think we should `require minitest/autorun` here, so it will fix the issue where there is a double run.

**Review by arthurnn — COMMENTED** [review]
I like this. I think we should indeed move back to autorun.. Thanks

File: railties/lib/rails/commands/test.rb (hunk 4 above)
Comment (grosser) [inline railties/lib/rails/commands/test.rb]: this will also fix the double run since autorun guards against being called twice, but it's simpler to do the require :)

**Comment by grosser** [comment]
any thought on making the test responsible for calling the autorun ? ... I'd guess there is a reason why there is so much logic in the code ...

**Comment by grosser** [comment]
tested single_cov and simple_cov and they both work with this ... single_cov will need a small patch, but I can take care of that ...

**Comment by arthurnn** [comment]
> any thought on making the test responsible for calling the autorun ? ... I'd guess there is a reason why there is so much logic in the code ...

Not sure I understand the question.

> tested single_cov and simple_cov and they both work with this ... single_cov will need a small patch, but I can take care of that ...

❤️

**Comment by arthurnn** [comment]
thanks

**Comment by grosser** [comment]
fixed in single_cov 0.5.3 ... https://github.com/grosser/single_cov/commit/7746a77590c814bd2448cf29735185e37cc23606

**Comment by kaspth** [comment]
Sweet! ❤️

**Comment by kaspth** [comment]
Backported this to 5-0-stable in: 3d5bd57

**Comment by senny** [comment]
@grosser thanks 💛
